swraid/mdcheck: add mdcheck timer for swraid device scrub

mfenniak opened the pull request:
###### Description of changes

Adds new options `hardware.raid.swraid.monitor ` which, when enabled, ports a systemd timer from the mdadm package to NixOS that runs regularly scheduled scrubs on swraid md* devices.

To support the new NixOS module, this PR adjusts the mdadm package to: (a) package the mdcheck script, (b) remove the upstream systemd units (mdstart_*, mdmonitor*) that are not functional in NixOS.

Questions/uncertainties with this (raised in https://discourse.nixos.org/t/nixos-mdadm-mdcheck/23716):

- I’ve named the configuration added hardware.raid.swraid.monitor – is this a good naming scheme?

- `enable` defaults to false – this maintains compatibility with existing NixOS systems, but, has some risks of being an unsafe configuration since regular raid scrubbing is a somewhat standard safety practice. Should this be true, false, or maybe true-if-stateVersion>="23.05" so that it becomes the default on new systems?

- I removed some existing systemd units that are packaged by the mdadm package, but aren’t usable because they don’t subst in the right mdcheck/mdadm paths, and instead recreated the units via systemd.timers and systemd.services to allow configurability. Is this a good direction, or, would it be better to use the upstream unit files and tweak them?

- This doesn’t run the mdmonitor systemd service which has a couple of functionalities; (a) it emails or runs a program on a device event, (b) it swaps spare devices between raid arrays if appropriate on the event of a failure. I didn’t want to include it because getting the functionality (b) requires enabling the functionality (a), which would require explicit configuration for “what to do”. Is leaving this out pretty reasonable? It could be enabled pretty easily by another module, or enhanced in the future.


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

mfenniak committed:
nixos/swraid.nix: init mdcheck swraid module

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nixos-mdadm-mdcheck/23716/4

peterhoeg commented:
Feel free to use anything from here:
```nix
{ config, lib, pkgs, ... }:

let
  cfg = config.toupstream.hardware.raid;

  inherit (lib) mkEnableOption mkIf mkMerge mkOption optional optionalAttrs types;
  inherit (import ../../../common/lib.nix { inherit config lib pkgs; }) mkBoolOption;

  diskPath = e:
    "/dev/disk/by-${e.idBy}/${e.name}";

  mdadmConf = ''
    MAILADDR = ${cfg.monitor.recipient}
    MAILFROM = no-reply@${config.networking.hostName}

  ''
  + "DEVICE " + lib.concatMapStringsSep " "
    (e: ''${diskPath e}${cfg.partitionSuffix}'')
    cfg.drives
  + "\n\n"
  + lib.concatMapStringsSep "\n"
    (e: "ARRAY /dev/${e.device} " + lib.concatStringsSep " " (lib.mapAttrsToList (k: v: "${k}=${v}") {
      metadata = e.version;
      UUID = e.uuid;
      name = config.networking.hostName + ":" + e.name;
    }))
    cfg.arrays;

  commonServiceConfig = {
    PrivateTmp = true;
    ProtectSystem = "strict";
    ProtectHome = "tmpfs";
  };

  enable = builtins.length cfg.drives > 0;
  doMonitor = builtins.length cfg.arrays > 0;

  arrayPath = array:
    "/sys/block/${array.device}/md";

in
{
  meta.maintainers = with lib.maintainers; [ peterhoeg ];

  options.toupstream.hardware.raid = {
    partitionSuffix = mkOption {
      description = "Partition suffix";
      type = types.str;
      default = "-part1";
    };

    earlyInit = mkOption {
      description = "Set up the RAID in initrd to ensure that the RAID is assembled with the proper name";
      type = types.bool;
      default = true;
    };

    openFirewall = mkBoolOption "Open the firewall for services that can be read remotely." true;

    temperature = {
      enable = mkBoolOption "Drive temperatures" true;

      preferHddtemp = mkBoolOption "Prefer hddtemp instead of drivetemp" false;
    };

    optimize = {
      enable = mkEnableOption "optimize software RAID";

      stripeCache = mkOption {
        description = "Device stripe cache size";
        type = types.ints.positive;
        default = 4096;
      };

      readAhead = mkOption {
        description = "Device read-ahead";
        type = types.ints.positive;
        default = 65535;
      };

      speed = mkOption {
        description = "Maximum resync/rebuild speed";
        type = types.ints.positive;
        default = 500000;
      };
    };

    drives = mkOption {
      description = "List of drives for software RAID";
      default = [ ];
      type = types.listOf
        (types.submodule {
          options = {
            name = mkOption {
              description = "Drive name without the full path.";
              type = types.str;
            };

            idBy = mkOption {
              description = "The method to identify the drive.";
              type = types.enum [ "path" "wwn" ];
              default = "path";
            };
          };
        });
    };

    monitor = {
      recipient = mkOption {
        description = "Notification recipient";
        type = types.str;
      };
    };

    scrub = {
      enable = mkEnableOption "automatic scrubbing";

      startAt = mkOption {
        description = "When to scrub";
        type = types.str;
        default = "weekly";
      };
    };

    arrays = mkOption {
      description = "software RAID arrays";
      default = [ ];
      type = types.listOf
        (types.submodule {
          options = {
            device = mkOption {
              description = "Device";
              type = types.str;
            };

            uuid = mkOption {
              description = "Device UUID";
              type = types.str;
            };

            name = mkOption {
              description = "Device name";
              type = types.str;
            };

            version = mkOption {
              description = "Version";
              type = types.enum [ "0.9" "1.2" ];
              default = "1.2";
            };
          };
        });
    };
  };

  config = mkMerge [
    (mkIf enable {
      hardware = {
        sata.timeout = {
          deciSeconds = 70;
          drives = map (e: { name = e.name; idBy = "path"; }) cfg.drives;
        };
      };

      systemd.services.mdmonitor.enable = false;
    })

    (mkIf doMonitor {
      boot.initrd.mdadmConf = mkIf cfg.earlyInit mdadmConf;
      environment.etc."mdadm/mdadm.conf".text = mdadmConf;
      systemd.services.mdadm = {
        description = "Monitor RAID";
        wantedBy = [ "multi-user.target" ];
        serviceConfig = commonServiceConfig // {
          Type = "exec";
          ExecStart = "${pkgs.mdadm}/bin/mdadm --monitor --scan";
        };
      };
    })

    (mkIf cfg.optimize.enable {
      boot.kernel.sysctl = {
        "dev.raid.speed_limit_max" = cfg.optimize.speed;
      };

      systemd.services = lib.listToAttrs (map
        (e: lib.nameValuePair
          "optimize-${e.device}"
          {
            description = "Optimize software RAID speed";
            wantedBy = [ "multi-user.target" ];
            unitConfig.ConditionPathExists = [
              "/dev/${e.device}"
              (arrayPath e)
            ];
            script = ''

              ###
              # /dev/${e.device}
              ###

              # read-ahead
              ${lib.getBin pkgs.util-linux}/bin/blockdev \
                --setra ${toString cfg.optimize.readAhead} \
                /dev/${e.device}

              # stripe cache size
              echo ${toString cfg.optimize.stripeCache} > \
                ${arrayPath e}/stripe_cache_size

              # use all resources for recovery
              echo max > \
                ${arrayPath e}/sync_max
            '';
            serviceConfig = commonServiceConfig // {
              Type = "oneshot";
              PrivateNetwork = true;
            };
          })
        cfg.arrays);
    })

    (mkIf cfg.scrub.enable {
      systemd.services = lib.listToAttrs (map
        (e: lib.nameValuePair
          "scrub-${e.device}"
          {
            description = "Scrub ${e.device}";
            inherit (cfg.scrub) startAt;
            unitConfig.ConditionPathExists = [
              "/dev/${e.device}"
              (arrayPath e)
            ];
            serviceConfig = commonServiceConfig // {
              Type = "oneshot";
              ExecStart = "${pkgs.runtimeShell} -c 'echo scrub > ${arrayPath e}/sync_action'";
              PrivateNetwork = true;
            };
          })
        cfg.arrays);
    })

    (mkIf cfg.temperature.enable {
      boot.kernelModules = optional (!cfg.temperature.preferHddtemp) "drivetemp";

      hardware.sensor.hddtemp = {
        enable = cfg.temperature.preferHddtemp;
        drives = map diskPath cfg.drives;
      };

      networking.firewall.allowedTCPPorts = [
      ]
      ++ optional (cfg.openFirewall && cfg.temperature.preferHddtemp) 7634;
    })
  ];
}
```

ElvishJerricco commented:
I'm not a fan of the name `hardware.raid.swraid.monitor` if only because it expands to "hardware raid software raid" :P Something like `services.mdmonitor` or `services.mdcheck` would probably be better.

It's also probably better to use the upstream service units and patch in the correct paths. You can still do something like `systemd.services.mdcheck_start.environment.MDADM_CHECK_DURATION = cfg.checkDuration`, because NixOS will add that as a drop-in unit, which should override the env variable from the upstream unit.

aanderse commented:
> It's also probably better to use the upstream service units and patch in the correct paths. You can still do something like `systemd.services.mdcheck_start.environment.MDADM_CHECK_DURATION = cfg.checkDuration`, because NixOS will add that as a drop-in unit, which should override the env variable from the upstream unit.

👍

mfenniak committed:
Reuse upstream mdcheck_* systemd units

mfenniak committed:
Rename module to services.mdcheck

mfenniak commented:
@ElvishJerricco  @aanderse Thanks for the feedback, guys.  I wasn't aware of the drop-in unit concept -- I've adjusted the implementation to now use the upstream systemd units (patched), and renamed the module to `services.mdcheck`.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1544

SuperSandro2000 commented on nixos/modules/services/hardware/mdcheck.nix:
```diff
@@ -0,0 +1,62 @@
+  options.services.mdcheck = {
+    enable = mkEnableOption (lib.mdDoc ''
+      This option enables mdcheck timers that run regular full scrubs of the
+      md devices.  These processes can also cause high I/O utilization; the
```
```suggestion
      md devices. These processes can also cause high I/O utilization; the
```

SuperSandro2000 commented on nixos/modules/services/hardware/mdcheck.nix:
```diff
@@ -0,0 +1,62 @@
+    enable = mkEnableOption (lib.mdDoc ''
+      This option enables mdcheck timers that run regular full scrubs of the
+      md devices.  These processes can also cause high I/O utilization; the
+      configuration option `boot.kernel.sysctl."dev.raid.speed_limit_max"`
+      can be used to limit the I/O utilization of the mdcheck processes.
```
We should do a nix option for that

ElvishJerricco replied:
@SuperSandro2000 Why? It would just be trading one oneliner for another.

SuperSandro2000 replied:
To make it discoverable. If we already mention a particular setting in a description than this is a nice addition.

SuperSandro2000 commented on nixos/modules/services/hardware/mdcheck.nix:
```diff
@@ -0,0 +1,62 @@
+      default = "6 hours";
+      description = lib.mdDoc ''
+        When mdcheck starts to execute, this option controls how long should
+        be spent on the scrub operation before stopping.  The default
```
```suggestion
        be spent on the scrub operation before stopping. The default
```

SuperSandro2000 commented on nixos/modules/services/hardware/mdcheck.nix:
```diff
@@ -0,0 +1,62 @@
+      default = "1:05:00";
+      description = lib.mdDoc ''
+        systemd OnCalendar expression for when to continue an in-progress
+        mdcheck operation that was paused after `duration` time passed.  The
+        default is to start at 1:05am every day.  If there is no in-progress
+        operation then nothing will be started.
```
```suggestion
        mdcheck operation that was paused after `duration` time passed.
        The default is to start at 1:05am every day.
        If there is no in-progress operation then nothing will be started.
```

SuperSandro2000 commented on pkgs/os-specific/linux/mdadm/default.nix:
```diff
@@ -38,6 +38,17 @@ stdenv.mkDerivation rec {
+  postInstall = ''
+    install -D -m 755 misc/mdcheck ''${out}/bin/mdcheck
+    wrapProgram $out/bin/mdcheck \
+      --prefix PATH : ''${out}/bin:${lib.makeBinPath [ coreutils util-linux findutils gnugrep ]}
+
+    sed -i \
+      -e "s@/usr/share/mdadm/mdcheck@''${out}/bin/mdcheck@" \
+      -e "s@ExecStartPre=-/usr/lib/mdadm/mdadm_env.sh@@" \
+      ''${out}/lib/systemd/system/mdcheck* ''${out}/lib/systemd/system/mdmonitor*
```
```suggestion
      --prefix PATH : $out/bin:${lib.makeBinPath [ coreutils util-linux findutils gnugrep ]}

    sed -i \
      -e "s@/usr/share/mdadm/mdcheck@$out/bin/mdcheck@" \
      -e "s@ExecStartPre=-/usr/lib/mdadm/mdadm_env.sh@@" \
      $out/lib/systemd/system/mdcheck* $out/lib/systemd/system/mdmonitor*
```

If the module adds the mdadm package globally we don't need to prefix $out/bin

ElvishJerricco commented on nixos/modules/services/hardware/mdcheck.nix:
```diff
@@ -0,0 +1,62 @@
+      type = types.str;
+      default = "Sun *-*-1..7 1:00:00";
+      description = lib.mdDoc ''
+        systemd OnCalendar expression for when to start an mdcheck operation.
+        The default is to begin on the first Sunday of every month at 1am.
```
The doc to link to would be the `CALENDAR EVENTS` section of `systemd.time(7)`.

A random delay can be accomplished with `RandomizedDelaySec`, but anyway, I think for consistency's sake, we should probably have this implemented equivalently to the `services.zfs.autoScrub`, which defaults to `Sun, 02:00` and no random delay. (Though I've just noticed that `services.btrfs.autoScrub` has a *different* default of `monthly`, which I find disappointing).

ctheune commented:
This has been sitting here for a while and I just recently was in the weeds with #123466 and #254429.

After reviewing the upstream units carefully I strongly vote against getting rid of them. The upstream udev rules and systemd units are intricately intertwined and rebuilding them on the NixOS side will quickly turn into an unsustainable hot mess. Looking at the unreleased changes upstream.

If the check timer really doesn't activate (I didn't check that, I only worked to fix the monitor service in those previous PRs), then I'm all for using the drop-in unit approach and keeping the changeset as minimal as possible but would strongly recommend to extend the test. I've broken master yesterday which warranted a further extension of the tests and as this is storage, we need to be careful not having to say "sorry" to our users by screwing up currently working systems.

ElvishJerricco commented:
@ctheune I believe the PR is currently intended to use the upstream units; it's just currently missing the line `systemd.packages = [ pkgs.mdadm ];`, which is necessary before merging.

ulysses4ever commented:
> renamed the module to `services.mdcheck`

This is way worse for discoverability than anything that has `raid` in it.

ctheune commented:
I think this PR isn't needed anymore anyway. We should double check that the original upstream modules really do start the timer, though.

lierdakil commented:
See https://github.com/NixOS/nixpkgs/issues/332096, TL;DR: the systemd services installed from the upstream are completely non-functional as they hard-code a path to a script in `/usr/share` that isn't installed in the first place (and, actually, shouldn't be installed at that path on NixOS anyway). TL;DR: automatic scrubbing is currently entirely non-functional on NixOS. Which is mildly to very bad, depending on how you look at it.

ctheune commented:
Dang. I'd love for the situation to improve in general but I'm a bit puzzled what a good strategy is to not end up in maintenance hell whenever upstream decides to fiddle around in this (excuse me) rat's nest of udev rules and scripts ... 

Ideas? Happy to talk on Element about this.

lierdakil commented:
> rat's nest of udev rules and scripts

Well, the _good news_ is automatic scrubbing is just two systemd timer units and one bash script. Nothing complicated or particularly convoluted there. The straightforward approach is to install the bash script to the nix store and patch systemd services with the correct path. Or we could reimplement those units on the Nix side, they're extremely simple. Both are viable IMO, with different tradeoffs. Personally, I'm leaning to the former, as if something changes, it'll be easier to catch at build time.

ctheune commented:
Yeah, definitely the former IMHO. 

I wonder whether I should go through all of those udev rules once to see what's going on there and whether patching/scrubbing has even a chance of getting it right for a long-ish period of time.

lierdakil commented:
> udev rules

They're not involved with the issue at hand (#332096). At all.

That being said, there are some references to non-existent systemd units in those rules, but that's an entirely separate concern.

ctheune commented:
Yeah, sorry for the tangent. I've touched that part before and it's really fragile, so I'm looking at the bigger picture. I'd support your suggested change for this specific issue.

mfenniak commented:
I'm no longer using mdraid, and won't complete any outstanding work on this PR.  If someone wants to adopt this change in the future and complete it, please feel free to use the code in this PR without restriction.  Closing.
